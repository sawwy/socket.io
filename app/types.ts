export type UserType = {
  isOnline: boolean;
  lastSeen: Date;
  joined: Date;
  username: string;
  avatar: string;
  description: string;
};

export interface UserResponseType
  extends Omit<UserType, "lastSeen" | "joined"> {
  lastSeen: string;
  joined: string;
}

export type SelectedUserDataType = {
  user: UserType;
  rect: DOMRect;
};
