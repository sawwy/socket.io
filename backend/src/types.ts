import { MessageTypeEnum } from "./enums.ts";

export interface IUser {
  id: string;
  username: string;
  isOnline: boolean;
  lastSeen: Date;
  joined: Date;
}

export interface IUserMessage {
  messageType: MessageTypeEnum.UserMessage;
  message: string;
  user: IUser;
  timestamp: string;
}

export interface IUsers {
  [userId: string]: IUser;
}
