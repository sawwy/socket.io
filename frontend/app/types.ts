export interface IUser {
  id: string;
  isOnline: boolean;
  lastSeen: Date;
  joined: Date;
  username: string;
  avatar?: string;
  description?: string;
}

export interface UserResponseType extends Omit<IUser, "lastSeen" | "joined"> {
  lastSeen: string;
  joined: string;
}

export type SelectedUserDataType = {
  user: IUser;
  rect: DOMRect;
};

export interface IMessage {
  message: string;
  username: string;
  timestamp: string;
}
