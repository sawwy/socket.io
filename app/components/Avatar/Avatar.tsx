import { UserType } from "~/types";
import styles from "./styles.module.css";

type AvatarPropsType = {
  user: UserType;
};

export const Avatar = ({ user }: AvatarPropsType) => {
  const idleTreshold = new Date();
  idleTreshold.setMinutes(idleTreshold.getMinutes() - 15);

  const isIdle = user.lastSeen.getTime() - idleTreshold.getTime() < 0;

  return (
    <div className={styles.avatar}>
      <img alt={`avatar of user ${user.username}`} src={user.avatar} />
      {user.isOnline && (
        <div className={styles.status}>
          <div className={styles.online} />
          {isIdle && <div className={styles.idle} />}
        </div>
      )}
    </div>
  );
};
