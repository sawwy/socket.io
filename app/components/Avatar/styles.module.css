.avatar {
  position: relative;
  width: 32px;

  img {
    width: 32px;
    border-radius: 16px;
  }
}

.status {
  display: grid;
  position: absolute;
  bottom: 0;
  right: 0;
  background-color: rgb(32, 34, 37);
  height: 0.7rem;
  width: 0.7rem;
  border-radius: 0.35rem;
}

.online {
  display: flex;
  position: relative;
  background-color: rgb(33, 152, 71);
  height: 0.5rem;
  width: 0.5rem;
  border-radius: 0.25rem;
  align-self: center;
  justify-self: center;
}

.idle {
  position: absolute;
  display: flex;
  background-color: rgb(32, 34, 37);
  height: 6px;
  width: 6px;
  border-radius: 3px;
  top: 1px;
  left: 1px;
}
