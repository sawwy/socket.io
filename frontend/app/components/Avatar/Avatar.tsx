import { IUser } from "~/types";
import styles from "./styles.module.css";
import globalStyles from "~/styles/styles.module.css";
import { getColorHighlight } from "~/utils/colorUtils";

type AvatarPropsType = {
  user: IUser;
};

export const Avatar = ({ user }: AvatarPropsType) => {
  const idleTreshold = new Date();
  idleTreshold.setMinutes(idleTreshold.getMinutes() - 15);

  const isIdle = user.lastSeen.getTime() - idleTreshold.getTime() < 0;
  const usernameAbbreviation = user.username.slice(0, 2);

  return (
    <div className={styles.avatar}>
      <div
        className={`${styles.avatarContent} ${getColorHighlight(
          user,
          globalStyles
        )}
        )}`}
        aria-label={`avatar of user ${user.username}`}
      >
        <div>{usernameAbbreviation}</div>
      </div>
      {user.isOnline && (
        <div className={styles.status}>
          <div className={styles.online} />
          {isIdle && <div className={styles.idle} />}
        </div>
      )}
    </div>
  );
};
