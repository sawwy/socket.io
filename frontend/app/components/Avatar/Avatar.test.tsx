import renderer from "react-test-renderer";
import { Avatar } from "./Avatar";
import { IUser } from "~/types";

const testUser: IUser = {
  id: "1234-asdf",
  isOnline: true,
  lastSeen: new Date("2024-04-02T16:00:00"),
  joined: new Date("2023-03-01T08:00:00"),
  username: "qwerty",
};

describe("Avatar Snapshot", () => {
  it("Matches DOM Snapshot", () => {
    const domTree = renderer.create(<Avatar user={testUser} />).toJSON();
    expect(domTree).toMatchSnapshot();
  });
});
