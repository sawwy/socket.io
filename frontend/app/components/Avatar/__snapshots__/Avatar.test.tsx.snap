// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar Snapshot Matches DOM Snapshot 1`] = `
<div
  className="avatar"
>
  <div
    aria-label="avatar of user qwerty"
    className="avatarContent colorFillGreen
        )}"
  >
    <div>
      qw
    </div>
  </div>
  <div
    className="status"
  >
    <div
      className="online"
    />
    <div
      className="idle"
    />
  </div>
</div>
`;
